[EP] Agent Gateway Integration (#11152)
kgateway-dev/kgateway: design/11150.md

       }
     ],
     "targets": {
       "mcp": [
         {
           "name": "everything",
           "stdio": {
             "cmd": "npx",
             "args": [
               "@modelcontextprotocol/server-everything"
             ]
           }
         }
       ]
     }
   }
```

kgateway will need to add support for syncing the configuration for AgentGateway via the xDS interface, along with any additional
APIs that kgateway needs to support AgentGateway.

## Motivation

As agent-based architectures grow more complex, there's a growing need to centralize the management of networking, 
security, and observability across diverse deployments. AgentGateway provides a unified data plane purpose-built for 
AI connectivity, with first-class support for protocols like MCP and A2A. 

By integrating kgateway with the AgentGateway, users can build and operate sophisticated agent ecosystems while 
cleanly separating infrastructure concerns from agent and tool logic. This enables secure, observable, and scalable 
connectivity for agents and tools running in Kubernetes.

Instead of natively supporting MCP and A2A in Envoy, integrating kgateway with AgentGateway allows us to offload 

> protocol-specific logic to a purpose-built control plane while still leveraging Envoy’s mature traffic routing, 

Reviewer: ```suggestion
protocol-specific logic to a purpose-built data plane while still leveraging Envoy’s mature traffic routing, 
```

security, and observability features. 

This separation of concerns ensures that Envoy remains a general-purpose, protocol-agnostic proxy, while AgentGateway 
handles agent-specific connectivity patterns. It also allows faster iteration on agent-native features without modifying

> core proxy infrastructure.

Reviewer: This is motivation for the data-plane, not the feature in the control-plane

Author: Yeah, I want to highlight the "we don't want to implement this in the Envoy proxy" part somehow. How about:

This separation of concerns allows us to use Envoy as a general-purpose, protocol-agnostic proxy, while AgentGateway will handle agent-specific connectivity patterns. It also allows faster iteration on agent-native features without modifying core proxy infrastructure.

The kgateway control plane will configure both Envoy and AgentGateway, enabling unified management while preserving the flexibility to evolve each data plane independently. This avoids entangling agent-specific logic in Envoy configuration while benefiting from AgentGateway's quick iteration cycles on new agent-specific features.

### Goals

> - Enable users to deploy kgateway to configure AgentGateway as a data plane.

Reviewer: ```suggestion
- Enable users to deploy kgateway to configure agentgateway as a data plane.
```

I think

Author: Okay I'll make it consistent!

## Implementation Details

An earlier MCP-specific implementation was introduced in kgateway [PR #11034](https://github.com/kgateway-dev/kgateway/pull/11034).
This EP will extend that support to other AgentGateway features (e.g. A2A, MCP multiplexing).

### Configuration
- Add helm option to enable/disable kgateway and AgentGateway integration
- Add support for MCP and A2A protocol configuration to kgateway's Gateway resource
- Add support for kgateway to configure authorization for AgentGateway.

#### Gateway protocol support

First a user will need to create a `GatewayParameters` and `GatewayClass` for kgateway to use to configure AgentGateway. 
The `GatewayParameters` resource that allows a user to set up self-managed Gateways. The `GatewayParameters` resource in this case
instructs kgateway to not automatically spin up a gateway proxy, but instead wait for the custom gateway proxy deployment.
The `GatewayClass` will need to bind it to the custom GatewayParameters.

```yaml
kind: GatewayParameters
apiVersion: gateway.kgateway.dev/v1alpha1
metadata:
  name: kgateway
spec:
  selfManaged: {}
---
kind: GatewayClass
apiVersion: gateway.networking.k8s.io/v1
metadata:

>   name: mcp

Reviewer: should this now be `agent`?

Author: Yep, actually it's `agentgateway` (But I'm open to changing it to `kgateway-agentgateway` to match waypoints)

        from: All
```

Then the user can define a Kubernetes-native Service:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: mcp
  namespace: gwtest
  labels:
    app: mcp
spec:
  clusterIP: "10.0.0.11"
  ports:
    - name: http
      port: 8080
      targetPort: 8080
      appProtocol: kgateway.dev/mcp
  selector:
    app: mcp
```

#### AgentGatewayAuthPolicy

One option is to introduce a new AuthPolicy for kgateway to configure the AgentGateway authentication. The Agent 
Gateway supports[JWT Authentication](https://agentgateway.dev/docs/security/jwt/) and [RBAC Authorization](https://agentgateway.dev/docs/security/rbac/) based on the JWT.

For example, the following AuthPolicy will allow the JWT token with claim `sub:me` to access the `convert_to_markdown` tool on the MCP server:

```yaml
apiVersion: gateway.kgateway.dev/v1alpha1
kind: AgentGatewayAuthPolicy
metadata:
  name: agent-gw-jq-auth
spec:
  targetRefs:
  - name: mcp-gateway
    group: gateway.networking.k8s.io
    kind: Gateway
  rules:
    - matches:
      - type: JWT
        jwt:
          claim: sub
          value: me
      resource:
        kind: tool
        name: default/mcp-tool:convert_to_markdown
```

> Another option is to introduce a generic authentication policy that kgateway can configure to handle authentication

Reviewer: Personally I'm much more in favor of this option

Author: @lgadban is too 😄

configuration for Envoy or kgateway. 

```
apiVersion: gateway.kgateway.dev/v1alpha1
kind: RequestAuthentication
metadata:
  name: "jwt-example"
  namespace: foo
spec:
  targetRefs:
  - name: mcp-gateway
    group: gateway.networking.k8s.io
    kind: Gateway
  jwtRules:
  - issuer: "testing@secure.kgateway.io"
    jwksUri: "https://raw.githubusercontent.com/istio/istio/release-1.25/security/tools/jwt/samples/jwks.json"
```

However, this approach would require targetRefs or filtering on specific kinds of resources (tools, agents, etc.). 
One option is to allow the `targetRefs` to not only apply to Gateways, but also to other resources (a2a services, mcp services, tool definitions, etc.).

> The `kind` field would be used to determine which resources to apply the policy to (AgentGateway listeners, AgentGateway targets, etc.).

Reviewer 2: so there's the authN part that can be common - configure issuer, and requiring a valid jwt (maybe this can go on the GatewayExtension?).
the authZ policy applies to routes. it would be nice if we can come up with a policy for both http and mcp/a2a

in the past we had jwt for authN and rbac policy for authZ. so i think the JWT policy can apply to both http and mcp.
rbac potentially also, as it matches principal (represented by a jwt token), with premissions. so i think we can potentially add tool permissions there.

Author: Yep, I was thinking of this separate EP for the JWT policies we need: https://github.com/kgateway-dev/kgateway/pull/11194 (So this enhancement relies on the JWT policy design to be resolved- either a unified policy for AgentGateway and Envoy, or a separate AgentGateway specific policy)

instruct kgateway to not automatically spin up a gateway proxy, but instead wait for your custom gateway proxy deployment.

The second phase of this EP will allow kgateway to manage the AgentGateway deployment. This will require addition Helm values
to configure the AgentGateway deployment (e.g., `.Values.agentGateway.deployment.version`, `.Values.agentGateway.deployment.image`). The 
deployment management mode (`selfManaged`) should work for either Envoy or AgentGateway deployments.

### Translator and Proxy Syncer
A new syncer should be added to kgateway to translate Gateway configuration into AgentGateway resources. This syncer will 
be responsible for converting relevant Gateway definitions into AgentGateway-compatible Listener and Target configurations.

The syncer will need to handle the new protocols (e.g., `kgateway.dev/mcp`, `kgateway.dev/a2a`). The `allowedRoutes` field
on the `Gateway` resource can be used to determine which namespaces will be selected for the AgentGateway. This allows target
discovery to be restricted to specific namespaces if desired.

### Plugin

> No plugin changes will be required as the AgentGateway syncer will be responsible for syncing the configuration independently of the plugin system.

Reviewer 2: more than that - because this is not envoy, none of our existing plugins can be used.

# EP-11150: agentgateway Support

* Issue: [#11150](https://github.com/kgateway-dev/kgateway/issues/11150)

## Background

> Integrate kgateway with the [agentgateway](https://github.com/agentgateway/agentgateway) as a data plane that provides AI connectivity for agents and tools in any environment.

Reviewer 3: nit: this line should mention kgateway being the control-plane for agentgateway

- Native integration of kgateway with the agentgateway's [observability features](https://agentgateway.dev/docs/observability/) 
such as metrics and tracing are in scope for this initial EP. These features should work alongside the agentgateway proxy out of the box, but kgateway will not expose any functionality to configure these features separately.
- Supporting configuring an OpenAPI-based MCP server via kgateway for agent and tool discovery is out of the initial scope. 
Allowing the user to configure an OpenAPI server as an MCP server is out of the initial scope, but would be a useful future extension. 
This would allow kgateway to serve an OpenAPI-compatible interface for tool discovery and interaction. The approach should be flexible to support this in the future if a user specifies a `openapi-mcp` type protocol.

## Implementation Details

An earlier MCP-specific implementation was introduced in kgateway [PR #11034](https://github.com/kgateway-dev/kgateway/pull/11034).
This EP will extend that support to other agentgateway features (e.g. A2A, MCP multiplexing).

### Configuration
- Add helm option to enable/disable kgateway and agentgateway integration

> - Add support for MCP and A2A protocol configuration to kgateway's Gateway resource

Reviewer 3: What is `kgateway's Gateway resource` referring to?

Author: The Kubernetes Gateway API resource. Will clarify!

- Add support for kgateway to configure authorization for agentgateway.

#### Gateway protocol support

First a user will need to create a `GatewayParameters` and `GatewayClass` for kgateway to use to configure agentgateway. 

> The `GatewayParameters` resource that allows a user to set up self-managed Gateways. The `GatewayParameters` resource in this case
> instructs kgateway to not automatically spin up a gateway proxy, but instead wait for the custom gateway proxy deployment.

Reviewer 3: ```suggestion
The `GatewayParameters` resource that allows a user to set up self-managed Gateways, that
instructs kgateway to not automatically spin up a gateway proxy, but instead wait for the custom gateway proxy deployment.
```

The `GatewayClass` will need to bind it to the custom GatewayParameters.

```yaml
kind: GatewayParameters
apiVersion: gateway.kgateway.dev/v1alpha1
metadata:
  name: kgateway
spec:
  selfManaged: {}
---
kind: GatewayClass
apiVersion: gateway.networking.k8s.io/v1
metadata:
  name: mcp
spec:
  controllerName: kgateway.dev/kgateway
  parametersRef:
    group: gateway.kgateway.dev
    kind: GatewayParameters
    name: kgateway
    namespace: default
```

> The `Gateway` resource will need be configured to use the custom `GatewayClass`. The custom `GatewayClass` instructs 

Reviewer 3: ```suggestion
The `Gateway` resource needs to be configured to use the custom `GatewayClass`. The custom `GatewayClass` instructs 
```

For example, you can open up an MCP listener on port 8080:

```yaml
kind: Gateway
apiVersion: gateway.networking.k8s.io/v1
metadata:
  name: mcp-gateway
spec:
  gatewayClassName: mcp
  listeners:
  - protocol: kgateway.dev/mcp
    port: 8080
    name: http
    allowedRoutes:
      namespaces:
        from: All
```

Then the user can define a Kubernetes-native Service:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: mcp
  namespace: gwtest
  labels:
    app: mcp
spec:
  clusterIP: "10.0.0.11"
  ports:
    - name: http
      port: 8080
      targetPort: 8080
      appProtocol: kgateway.dev/mcp
  selector:
    app: mcp
```

#### agentgatewayAuthPolicy

One option is to introduce a new AuthPolicy for kgateway to configure the agentgateway authentication. The Agent 
Gateway supports[JWT Authentication](https://agentgateway.dev/docs/security/jwt/) and [RBAC Authorization](https://agentgateway.dev/docs/security/rbac/) based on the JWT.

For example, the following AuthPolicy will allow the JWT token with claim `sub:me` to access the `convert_to_markdown` tool on the MCP server:

```yaml
apiVersion: gateway.kgateway.dev/v1alpha1

> kind: agentgatewayAuthPolicy

Reviewer 3: PascalCase naming.

Also weird to prefix the Kind with AgentGateway when the API group could be `agentgateway`. Do we have any common policies for regular and agentgateway?
